ansible_collections/arista/avd/roles/eos_validate_state/tasks/interface_state.yml

```diff
@@ -0,0 +1,81 @@
+---
+
+# Interface State Tests
+
+- name: Gather interfaces state
+  eos_command:
+    commands: "show interfaces description | json"
+  ignore_errors: "{{ validation_mode_loose }}"
+  register: interfaces_state
+  tags:
+    - interfaces_state
+
+- name: Validate Ethernet interfaces state
+  assert:
+    that:
+      - interfaces_state.stdout[0].interfaceDescriptions[ethernet_interface.key].interfaceStatus == "up"
+      - interfaces_state.stdout[0].interfaceDescriptions[ethernet_interface.key].lineProtocolStatus == "up"
+    quiet: true
+  loop: "{{ ethernet_interfaces | default({}) | dict2items }}"
+  loop_control:
+    loop_var: ethernet_interface
+  ignore_errors: "{{ validation_mode_loose }}"
+  register: ethernet_interface_state_results
+  tags:
+    - interfaces_state
+
+- name: Validate Port-Channel interfaces state
+  assert:
+    that:
+      - interfaces_state.stdout[0].interfaceDescriptions[port_channel_interface.key].interfaceStatus == "up"
+      - interfaces_state.stdout[0].interfaceDescriptions[port_channel_interface.key].lineProtocolStatus == "up"
+    quiet: true
+  loop: "{{ port_channel_interfaces | default({}) | dict2items }}"
+  loop_control:
+    loop_var: port_channel_interface
+  ignore_errors: "{{ validation_mode_loose }}"
+  register: port_channel_interface_results
+  tags:
+    - interfaces_state
+
+- name: Validate Vlan interfaces state
+  assert:
+    that:
+      - interfaces_state.stdout[0].interfaceDescriptions[vlan_interface.key].interfaceStatus == "up"
+      - interfaces_state.stdout[0].interfaceDescriptions[vlan_interface.key].lineProtocolStatus == "up"
+    quiet: true
+  loop: "{{ vlan_interfaces | default({}) | dict2items }}"
+  loop_control:
+    loop_var: vlan_interface
+  ignore_errors: "{{ validation_mode_loose }}"
+  when: (vlan_interfaces is defined and vlan_interfaces is not none)
```

Comment:
In case there is no SVI configured on the device, task generates error

```
TASK [eos_validate_state : Validate Vlan interfaces state] *************
Friday 07 August 2020  11:47:16 +0200 (0:00:01.696)       0:00:19.003 *********
Friday 07 August 2020  11:47:16 +0200 (0:00:01.696)       0:00:19.003 *********
fatal: [DC1-L2LEAF1A]: FAILED! =>
  msg: dict2items requires a dictionary, got <class 'NoneType'> instead.
...ignoring
```

The same error appears for port-channel: use case is leaf not running MLAG

Trying to find a workaround

Reply:
Yes, I experience the same error, this is because when combining Conditionals with a loop, the when: statement is processed separately for each item, so this is why we are seeing this error... https://docs.ansible.com/ansible/latest/user_guide/playbooks_loops.html

I believe we can resolve by introducing a block, and applying conditional on the block.
https://docs.ansible.com/ansible/latest/user_guide/playbooks_blocks.html

Reply:
after testing and reading how the conditional applies, to a block, the block strategy will not work either.
"The directive affects the block itself, but is inherited by the tasks enclosed by a block. i.e. a when will be applied to the tasks, not the block itself"

Reply:
Indeed. I am investigating with `set_fact` to see if we can do something to create a default empy structure for l2leafs

Reply:
Only solution I found was to include tasks_file with a conditional

```yaml
- name: Build fact
  set_fact:
    temp_asset: yes
  when: (vlan_interfaces is defined and vlan_interfaces is not none and vlan_interfaces|length > 1)
  tags:
    - interfaces_state

- name: Execute SVI check
  include_tasks: "temp.yml"
  when: (temp_asset is defined and temp_asset==true)
  tags:
    - interfaces_state
```

Not optimal at all, but seems to work

Reply:
I was able to make it work with set_fact like this:

```yaml
- name: Set vlan_interfaces to test
  set_fact:
    vlan_interfaces_test: "{{ vlan_interfaces }}"
    cacheable: yes
  when: (vlan_interfaces is defined and vlan_interfaces is not none)
  tags:
    - interfaces_state

- name: Validate Vlan interfaces state
  assert:
    that:
      - interfaces_state.stdout[0].interfaceDescriptions[vlan_interface.key].interfaceStatus == "up"
      - interfaces_state.stdout[0].interfaceDescriptions[vlan_interface.key].lineProtocolStatus == "up"
    quiet: true
  loop: "{{ vlan_interfaces_test | default({}) | dict2items }}"
  loop_control:
    loop_var: vlan_interface
  ignore_errors: "{{ validation_mode_loose }}"
  register: vlan_interface_state_results
  tags:
    - interfaces_state
```

Note I tried first to set var to an empty dictionary but was not successful.

yes, include_task is another option but will create more separate files.

What do you think?

Reply:
Your approach is far more efficient: no additional files and easy to troubleshoot.
We should apply same approach for Port Channel to cover non-mlag leaves

Reply:
ok, found another way with default filter set like this. ` | default({}, true)`
This set value when variable evaluates to empty.
See commit 15790d1